### DISCUSSION TITLE
The global operator override `new` looks unsafe.

### DISCUSSION THREAD

**Discussion by AndreaCatania** [jrouwe/JoltPhysics#571] [General]
Hello, while integrating a library based on the Jolt Physics; I've noticed that it uses `new` to create the `Body` objects https://github.com/jrouwe/JoltPhysics/blob/master/Jolt/Physics/Body/BodyManager.cpp#L148-L157.
Therefore, the `new` operator can be overridden to provide a custom allocator, to eventually support linear allocation.

The issue is that the global-override-`new` is not confined into the library, rather it s global per executable. It means that the rest of the application, where my library will be used, will also use the `new` defined by my library: this is unsafe and undesired.

I wonder if would be possible to switch to an Allocator solution, instead to rely on `new`.

Please, let me know if my understanding on how it works is wrong.

**Comment by mihe** [comment]
(I'm not a Jolt developer, just some random user of Jolt.)

Jolt overrides the `new` and `delete` operators per-class (i.e. not globally) using this macro:

https://github.com/jrouwe/JoltPhysics/blob/ad9cd403cc032bf56442803616d1081788df9d96/Jolt/Core/Memory.h#L28-L37

Which you'll find is used in `Body` here:

https://github.com/jrouwe/JoltPhysics/blob/ad9cd403cc032bf56442803616d1081788df9d96/Jolt/Physics/Body/Body.h#L32-L35

The functions that are used in this macro can be set by the user through these global function pointers:

https://github.com/jrouwe/JoltPhysics/blob/ad9cd403cc032bf56442803616d1081788df9d96/Jolt/Core/Memory.h#L19-L23

Which in examples like `HelloWorld` is typically done using this function, which defaults them to `malloc`/`free` and its aligned equivalents:

https://github.com/jrouwe/JoltPhysics/blob/ad9cd403cc032bf56442803616d1081788df9d96/Jolt/Core/Memory.cpp#L63-L69

You can find an example of me hooking up mimalloc to Jolt [here](https://github.com/godot-jolt/godot-jolt/blob/master/src/servers/jolt_globals.cpp).

**Reply by AndreaCatania** [reply]
Nice, this is much cleaner now, thx!
